executor/main.go

```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"os"
+	"time"
+
+	fluxhelmv2beta1 "github.com/fluxcd/helm-controller/api/v2beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/kubectl/pkg/scheme"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/polling"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/polling/aggregator"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/polling/collector"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/polling/event"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/status"
+	"sigs.k8s.io/cli-utils/pkg/object"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/yaml"
+)
+
+func main() {
+	var spec string
+	ctx := context.Background()
+
+	flag.StringVar(&spec, "spec", "", "The spec of the helmrelease object to apply")
+	flag.Parse()
+
+	if spec == "" {
+		os.Exit(1)
+	}
+
+	decodedSpec, err := base64.StdEncoding.DecodeString(spec)
+	if err != nil {
+		fmt.Printf("Failed to decode the string as a base64 string; got the string %v", spec)
+		os.Exit(1)
+	}
+
+	hr := &fluxhelmv2beta1.HelmRelease{}
+	yaml.Unmarshal(decodedSpec, hr)
+
+	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(clientcmd.NewDefaultClientConfigLoadingRules(), &clientcmd.ConfigOverrides{})
+	config, err := kubeConfig.ClientConfig()
+	if err != nil {
+		fmt.Printf("Failed to initialize the client config with %v", err)
+		os.Exit(1)
+	}
+	scheme := scheme.Scheme
+	if err := fluxhelmv2beta1.AddToScheme(scheme); err != nil {
+		fmt.Printf("Failed to add the flux helm scheme to the configuration scheme with %v", err)
+		os.Exit(1)
+	}
+	clientSet, err := client.New(config, client.Options{Scheme: scheme})
+	if err != nil {
+		fmt.Printf("Failed to create the clientset with the given config with %v", err)
+		os.Exit(1)
+	}
+
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: hr.Namespace,
+		},
+	}
+	// Best try at creating the namespace
+	clientSet.Create(ctx, ns)
+
+	if err := clientSet.Create(ctx, hr); err != nil {
+		fmt.Printf("Failed to create the helmrelease with %v", err)
+		os.Exit(1)
+	}
+
+	identifiers := object.ObjMetadata{
+		Namespace: hr.Namespace,
+		Name:      hr.Name,
+		GroupKind: schema.GroupKind{
+			Group: "helm.toolkit.fluxcd.io",
+			Kind:  "HelmRelease",
+		},
+	}
+
+	// We give the poller a minute before we time it out
+	if err := PollStatus(ctx, clientSet, config, time.Minute*2, identifiers); err != nil {
```

Comment:
Is this the duration it waits for the `HelmRelease` to succeed? We might want to make this configurable and the defaults longer.

Reply:
Yes, I agree; let's make this configurable. We can pass this through the application group spec

Reply:
I'm going to roll this up into the argo spec for now. I think we need more discussion on how this looks in the api. Additionally, the status timeout is almost meaningless right now because as soon as a failure condition is hit, the application controller triggers remediation and rollback